mobly/snippet/general_callback_handler.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2022 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A general handler class for Mobly Snippet Lib's callback events.
+
+The protocol for pulling events expected by this module is:
+
+RPC 1: eventWaitAndGet: Waits and returns an event for the specified identifier
+                        with timeout.
+JSON RPC Request:
+{
+  'id': <RPC-id>,
+  'method': 'eventWaitAndGet',
+  'params': (<callback_id: string>, <event_name: string>, <timeout_ms: int>),
+}
+
+RPC 2: eventGetAll: Gets all existing events for the specified identifier
+                    without waiting.
+JSON RPC Request:
+{
+  'id': <RPC-id>,
+  'method': 'eventGetAll',
+  'params': (<callback_id: string>, <event_name: string>),
+}
+
+Timeout error message:
+In the response of eventWaitAndGet, this class checks for the existence of
+the substring "EventSnippetException: timeout." in the error message. If it
+exists, this class throws a timeout error from the original error.
+"""
+from mobly.snippet import callback_handler_base
+from mobly.snippet import errors
+
+
+class GeneralCallbackHandler(callback_handler_base.CallbackHandlerBase):
```

Comment:
What makes this "general"?
This is general as opposed to what?
And why do we have both general handler and handler base...

This is quite confusing

Reply:
The Android and Windows clients use the general callback handler and the iOS client will use IosCallbackHandler. 

Use "General" here because we recommend future implementations of other clients to use this client when possible (this has restrictions on server-side RPC function names).

And the handler base is an abstract class which can not be used directly.

Does this make sense?

Reply:
The iOS one is special because it has limitation on the function name of RPCs, which are for pulling events from server.

Through writing in this way, I'm trying to make this point clearer: if you need to implement a new callback handler, which methods should you override. These methods are marked as abstractmethod and you cannot instantiate you class without overriding them. This seems clearer than reminding the user via docstring which functions should be overridden.

So it seems that I'm making things complex and confusing?